addons/mail/models/discuss/ir_websocket.py

```diff
@@ -79,28 +68,72 @@ def _build_bus_channel_list(self, channels):
         all_user_channels = self.env["discuss.channel"].search(domain)
         member_specific_channels = [(c, "members") for c in all_user_channels if c.id not in discuss_channel_ids]
-        channels.extend([*all_user_channels, *member_specific_channels])
+        channels.extend([*all_user_channels, *member_specific_channels, *presence_channels])
         return super()._build_bus_channel_list(channels)
 
+    def _prepare_subscribe_data(self, channels, last):
+        data = super()._prepare_subscribe_data(channels, last)
+        data["channels"].update(self._build_bus_channel_list(channels))
+        return data
+
+    def _dispatch_missed_presences(self, channels):
+        # There is a gap between a subscription client side (which is debounced)
+        # and the actual subcription thus presences can be missed. Send a
+        # notification to avoid missing presences during a subscription.
+        presence_channels = [c for c in channels if isinstance(c, tuple) and c[1] == "presence"]
+        partners = self.env["res.partner"].browse(
+            [p.id for p, _ in presence_channels if isinstance(p, self.pool["res.partner"])]
+        )
+        guest_ids = [g.id for g, _ in presence_channels if isinstance(g, self.pool["mail.guest"])]
+        domain = expression.AND(
+            [
+                [("last_poll", ">", datetime.now() - timedelta(seconds=2))],
+                expression.OR(
+                    [
+                        [("user_id", "in", partners.with_context(active_test=False).sudo().user_ids.ids)],
+                        [("guest_id", "in", guest_ids)]
+                    ]
+                ),
+            ]
+        )
+        # sudo: mail.presence: can access presences linked to presence channels.
+        missed_presences = self.env["mail.presence"].sudo().search(domain)
+        bus_target = (
+            self.env.user.partner_id
+            if self.env.user and not self.env.user._is_public()
+            else self.env["mail.guest"]._get_guest_from_context()
+        )
+        missed_presences._send_presence(bus_target=bus_target)
+
+    def _subscribe(self, og_data):
+        data = super()._subscribe(og_data)
+        self._dispatch_missed_presences(data["channels"])
+
     @add_guest_to_context
     def _update_bus_presence(self, inactivity_period):
         super()._update_bus_presence(inactivity_period)
-        if not self.env.user or self.env.user._is_public():
-            guest = self.env["mail.guest"]._get_guest_from_context()
-            if not guest:
-                return
-            # sudo: bus.presence - guests currently need sudo to write their own presence
-            self.env["bus.presence"].sudo().update_presence(
-                inactivity_period,
-                identity_field="guest_id",
-                identity_value=guest.id,
-            )
+        guest = self.env["mail.guest"]._get_guest_from_context()
+        if self.env.user._is_public() and not guest:
```

Comment:
We also assumed user or guest existed in many instances, but that said I'd rather not increase the diff any more and check for `self.env.user` existing instead of removing it from all other instances.

Reply:
We can just add the check here, and clean it later. However, I don't think we should assume the presence of an internal user or a guest. When I look at the `ir.websocket` methods and the mail overrides in the `master` branch, I don’t see evidence that such an assumption exists, nor do I think we should introduce it in this refactor.